packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1666,6 +1666,44 @@ function updateSyncExternalStore<T>(
   return nextSnapshot;
 }
 
+function updateSyncExternalStore<T>(
+  subscribe: (() => void) => () => void,
+  getSnapshot: () => T,
+  getServerSnapshot?: () => T,
+): T {
+  const hook = updateWorkInProgressHook();
+  const prevSnapshot = hook.memoizedState;
+  return updateSyncExternalStoreImpl(
+    hook,
+    prevSnapshot,
+    subscribe,
+    getSnapshot,
+    getServerSnapshot,
+  );
+}
+
+function rerenderSyncExternalStore<T>(
+  subscribe: (() => void) => () => void,
+  getSnapshot: () => T,
+  getServerSnapshot?: () => T,
+): T {
+  const hook = updateWorkInProgressHook();
+  const prevSnapshot =
+    currentHook === null
+      ? // This is a rerender during a mount.
+        hook.memoizedState
```

Comment:
Hmm I know this is equivalent to what #25650 did, but now that I look closer I think this is wrong for initial render. `prevSnapshot` should represent what's currently rendered on the screen. If there is no previous snapshot, then we should always schedule a commit effect. That's why mountSyncExternalStore always schedules an effect, without checking first.

So I think if `currentHook` is null, this should call `mountSyncExternalStore` instead.

Initial render should always mount an effect.

Reply:
Hm wouldn't this also mean we no longer use the client snapshot when we rerender during hydration?

Tried 
```js
function rerenderSyncExternalStore<T>(
  subscribe: (() => void) => () => void,
  getSnapshot: () => T,
  getServerSnapshot?: () => T,
): T {
  if (currentHook === null) {
    return mountSyncExternalStore(subscribe, getSnapshot, getServerSnapshot);
  } else {
    const hook = updateWorkInProgressHook();
    // This is a rerender during an update.
    return updateSyncExternalStoreImpl(
      hook,
      hook.memoizedState,
      subscribe,
      getSnapshot,
      getServerSnapshot,
    );
  }
}
```
and this leads to 
```
 FAIL  packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js
  ● Shared useSyncExternalStore behavior (shim and built-in) › regression test for #23150

    Rendered more hooks than during the previous render.
```

Reply:
Need to call `mountWorkInProgressHook` and then pass in as an argument. The rest of the logic is the same.